sphinx/source/docs/dev_guide/testing.rst

```diff
@@ -1,231 +1,447 @@
-can view at ``examples.log`` in the same directory that you the tests
-were run from.
+.. note::
+    Unit and integration tests require a recent version of Chrome or Chromium.
+    The BokehJS testing framework starts the browser automatically with the
+    right settings to guarantee consistent test results.
 
-Writing Tests
--------------
+.. _devguide_testing_local_typescript_devtools:
 
-In order to help keep Bokeh maintainable, all Pull Requests that touch code
-should normally be accompanied by relevant tests. While exceptions may be
-made for specific circumstances, the default assumption should be that a
-Pull Request without tests may not be merged.
+Testing with devtools server
+''''''''''''''''''''''''''''
 
-Python Unit Tests
-~~~~~~~~~~~~~~~~~
+In addition to running BokehJS tests from the command line, you can also use
+the BokehJS devtools server. This system requires the Chrome web browser to be
+available on your system. Use the BokehJS devtools server to run tests and
+review the visual tests' output.
 
-Python unit tests maintain the basic functionality of the Python portion of
-the Bokeh library. A few general guidelines will help you write Python unit
-tests:
+First, start the devtools server from the `bokehjs` subdirectory with the
+following command:
 
-absolute imports
-    In order to ensure that Bokeh's unit tests as relocatable and unambiguous
-    as possible, always prefer absolute imports in test files. When convenient,
-    import and use the entire module under test:
+.. code-block:: sh
 
-    * **GOOD**: ``import bokeh.models.transforms as bmt``
-    * **GOOD**: ``from bokeh.embed import components``
-    * **BAD**: ``from ..document import Document``
+    $ node test/devtools server
+    listening on 127.0.0.1:5777
 
-pytest
-    All new tests should use and assume `pytest`_ for test running, fixtures,
-    parameterized testing, etc. New tests should *not* use the ``unittest``
-    module of the Python standard library.
+You can now use the devtools server for the following operations:
 
-Integration Tests
-~~~~~~~~~~~~~~~~~
+Inspecting visual test results
+    After running integration tests, you can use the devtools server to
+    compare your local results with the baseline images. Open the displayed
+    server URL (usually ``127.0.0.1:5777``) in the Chrome web browser and append
+    ``/integration/report``. This will open a comparison view of any tests where
+    your locally rendered plot is different from the baseline file. For example:
+
+    .. image:: /_images/bokehjs_devtools_report.png
+        :class: image-border
+        :alt: Screenshot of devtool displaying a locally rendered image, an
+            image diff and a baseline image.
+
+Initiate test runs
+    You can also use the devtools server to initiate test runs. You have two
+    options:
+
+    * Run tests from a JavaScript console
+        Open one of these three endpoints in your web browser:
+
+        * ``/unit``
```

Comment:
@mattpap As mentioned earlier today, running `Tests.run_all()` at `127.0.0.1:5777/unit` on Ubuntu 20.04 (Chromium 92.0.4515.131) runs all the tests but also opens a new browser window with the URL `http://stuff.com/baz.html`. I find this a little confusing... In case this is not the intended behavior, I'm happy to open an issue about that!

![image](https://user-images.githubusercontent.com/39711796/128810449-c56b4d9f-c583-412f-ab40-c66d36277cbb.png)

Reply:
I can see now that navigating to `http://localhost:5777/unit/run?k=OpenURL` triggers popup blocker in my browser. We need to mock `OpenURL.navigate()` to avoid this. Please start an issue for this.

Reply:
Actually nevermind, I already created the issue and fixed it.

Reply:
Thanks!